File: leetcode/algorithms/441. 排列硬币/2.java
// 二分
class Solution {
    public int arrangeCoins(int n) {
       int low= 0;
       int high = n;
       while(low<=high){
           int mid = (high-low)/2+low;
           int total = mid*(mid+1)/2;
           if(total==n){
               return mid;
           }else if(total>n){
               high = mid-1;
           }else{
               low = mid+1;
           }
       }
       // 没找到的情况下
        return high;
    }
   
}

File: leetcode/剑指offer/剑指 Offer 11. 旋转数组的最小数字（调库了，可以试试二分）/2.cpp
class Solution {
public:
    int minArray(vector<int>& numbers) {
        int l =0;
        int r = numbers.size()-1;
        while(l<r){
            int mid = (r-l)/2+l;
            // 3 4 1 2 3
            if(numbers[mid]<numbers[r]){
                r= mid;
            //  2 3 4 5 1
            }else if (numbers[mid]>numbers[r]){
                l = mid+1;
            }else{
                // num[mid]  == num[r]
                r-=1;
            }
        }
        return numbers[l];
    }
};
